src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Diagnostics/RazorTranslateDiagnosticsService.cs

```diff
@@ -170,15 +171,53 @@ private Diagnostic[] FilterCSharpDiagnostics(Diagnostic[] unmappedDiagnostics, R
             return false;
         }
 
-        var owner = syntaxTree.GetOwner(sourceText, d.Range.End, logger);
+        var owner = syntaxTree.FindInnermostNode(sourceText, d.Range.End, logger);
         if (owner is null)
         {
             return false;
         }
 
-        var isCSharp = owner.Kind is SyntaxKind.CSharpExpressionLiteral or SyntaxKind.CSharpStatementLiteral or SyntaxKind.CSharpEphemeralTextLiteral;
+        if (IsCsharpKind(owner))
+        {
+            return true;
+        }
+
+        // If the owner is a markup text literal, the token found
+        // could still have been the end of a csharp statement. For example:
+        // CSS123456 gets reported with the following range
+        // <style> body { overflow: [|@DateTime.Now|]; } </style>
+        // FindInnermostNode will find the encompasing MarkupTextLiteral that ends with
+        // ';' but contains the @DateTime.Now csharp expression. Check for these cases
+        // by looking to see if the start is a transition before modifying the end position
+        // to be left one.
+        if (owner.Kind is SyntaxKind.MarkupTextLiteral)
+        {
+            if (!d.Range.Start.TryGetAbsoluteIndex(sourceText, logger, out var startIndex))
+            {
+                return false;
+            }
+
+            var startToken = syntaxTree.Root.FindToken(startIndex);
+            if (startToken.Kind is not SyntaxKind.Transition)
+            {
+                return false;
+            }
+
+            if (!d.Range.End.TryGetAbsoluteIndex(sourceText, logger, out var endIndex))
+            {
+                return false;
+            }
+
+            var newOwner = syntaxTree.Root.FindInnermostNode(endIndex - 1);
```

Comment:
The comment above says the `MarkupTextLiteral` "contains the csharp expression", so can we get it by navigating the tree (ie, the `Children` property I guess?) rather than walking to find the node again? I guess the same could apply to finding the transition at the start potentially, depending on what the tree looks like.

Reply:
Alternatively, could this use this whole bit use the new `FindNode` method that takes a span? That would at least give it two uses :)